src/Build/Graph/GraphBuilder.cs

```diff
@@ -237,79 +238,31 @@ void AddGraphBuildGlobalVariable(PropertyDictionary<ProjectPropertyInstance> glo
         ///     Maintain a queue of projects to be processed and evaluate projects in parallel
         ///     Returns false if loading the graph is not successful
         /// </summary>
-        private bool FindGraphNodes(
-            ConcurrentQueue<ConfigurationMetadata> projectsToEvaluate,
-            ProjectCollection projectCollection,
-            ProjectGraph.ProjectInstanceFactoryFunc projectInstanceFactory,
-            ProjectInterpretation projectInterpretation,
-            ConcurrentDictionary<ConfigurationMetadata, ProjectGraphNode> allParsedProjects,
-            out List<Exception> exceptions)
+        private Dictionary<ConfigurationMetadata, ProjectGraphNode> FindGraphNodes()
         {
-            var tasksInProgress = new ConcurrentDictionary<ConfigurationMetadata, object>();
-            var evaluationWaitHandle = new AutoResetEvent(false);
-            var exceptionsInTasks = new ConcurrentBag<Exception>();
-
-            while (projectsToEvaluate.Count != 0 || tasksInProgress.Count != 0)
+            foreach (ConfigurationMetadata projectToEvaluate in _entryPointConfigurationMetadata)
             {
-                ConfigurationMetadata projectToEvaluate;
-                if (projectsToEvaluate.Count != 0)
-                {
-                    projectToEvaluate = projectsToEvaluate.Dequeue();
-                    var task = new Task(
-                        () =>
-                        {
-                            var parsedProject = CreateNewNode(projectToEvaluate, projectCollection, projectInstanceFactory, allParsedProjects);
+                ParseProject(projectToEvaluate);
+            }
 
-                            foreach (var (referenceConfig, _) in projectInterpretation.GetReferences(parsedProject.ProjectInstance))
-                            {
-                                /*todo: fix the following double check-then-act concurrency bug: one thread can pass the two checks, loose context,
-                             meanwhile another thread passes the same checks with the same data and inserts its reference. The initial thread regains context
-                             and duplicates the information, leading to wasted work
-                             */
-                                if (!tasksInProgress.ContainsKey(referenceConfig))
-                                {
-                                    if (!allParsedProjects.ContainsKey(referenceConfig))
-                                    {
-                                        projectsToEvaluate.Enqueue(referenceConfig);
-                                        evaluationWaitHandle.Set();
-                                    }
-                                }
-                            }
-                        });
+            _graphWorkSet.WaitForAllWork();
+            _graphWorkSet.Complete();
+            _graphWorkSet.WaitForCompletion();
```

Comment:
I find this calling pattern confusing. Why is an explicit call to `WaitForAllWork` required? I think I understand that you have to `Complete()` to signal "no new incoming work", but the other two calls both feel like "wait for stuff" to me.

Reply:
I went ahead and merged these into a single call in #4508. It's a holdover from QB where we have two parse phases and multiple work sets which can enqueue work or wait against each other, so these were split into granular methods.